megatron/core/models/common/embeddings/rotary_pos_embedding.py

```diff
@@ -229,24 +235,146 @@ def apply_rotary_pos_emb(
     Reroute to the appropriate apply_rotary_pos_emb function depending on
     fused/unfused kernels, or bshd (conventional) / thd (packed seq) format
     """
-    if config.apply_rope_fusion and not HAVE_APPLY_ROPE_FUSION:
-        # setting apply_rope_fusion in config to False so that subsequent queries to this config also return False
-        config.apply_rope_fusion = False
-        if not getattr(apply_rotary_pos_emb, "printed_fused_warning", False):
-            logger.warning(
-                "Setting apply_rope_fusion to false because its implementation"
-                " is not included in Apex. Try upgrading to the latest version"
-            )
-            apply_rotary_pos_emb.printed_fused_warning = True
-    if config.apply_rope_fusion:
-        if cu_seqlens is None:
-            return fused_apply_rotary_pos_emb(t, freqs, transpose_output_memory=True)
-        else:
-            return fused_apply_rotary_pos_emb_thd(t, cu_seqlens, freqs)
+
+    if config.use_te_fused_rope and HAVE_TE and torch.cuda.is_available() and torch.version.hip:
+        return apply_rotary_pos_emb_fused_te(t = t, freqs = freqs, config = config, cu_seqlens = cu_seqlens)
     else:
-        if cu_seqlens is None:
-            return apply_rotary_pos_emb_bshd(t, freqs, rotary_interleaved=config.rotary_interleaved)
+        if config.apply_rope_fusion and not HAVE_APPLY_ROPE_FUSION:
+            # setting apply_rope_fusion in config to False so that subsequent queries to this config also return False
+            config.apply_rope_fusion = False
+            if not getattr(apply_rotary_pos_emb, "printed_fused_warning", False):
+                logger.warning(
+                    "Setting apply_rope_fusion to false because its implementation"
+                    " is not included in Apex. Try upgrading to the latest version"
+                )
+                apply_rotary_pos_emb.printed_fused_warning = True
+
+        if config.apply_rope_fusion:
+            if cu_seqlens is None:
+                return fused_apply_rotary_pos_emb(t, freqs, transpose_output_memory=True)
+            else:
+                return fused_apply_rotary_pos_emb_thd(t, cu_seqlens, freqs)
         else:
-            return apply_rotary_pos_emb_thd(
-                t, cu_seqlens, freqs, rotary_interleaved=config.rotary_interleaved
-            )
+            if cu_seqlens is None:
+                return apply_rotary_pos_emb_bshd(t, freqs, rotary_interleaved=config.rotary_interleaved)
+            else:
+                return apply_rotary_pos_emb_thd(
+                    t, cu_seqlens, freqs, rotary_interleaved=config.rotary_interleaved
+                )
+
+
+class FusedRoPEFunc(torch.autograd.Function):
```

Comment:
I should have given this comment earlier. I think it is better to move this to rope_utils.py. In upstream, they figure out if the Apex kernel is there or not and decide whether to use fused RoPE or not (https://github.com/NVIDIA/Megatron-LM/blob/772faca1f8d5030621b738cbd8e8bb2d8d28f6e6/megatron/core/models/common/embeddings/rope_utils.py#L19-L27) . We can do similar things there with the TE kernels. I think it would aligns with upstream better and would make it easier for future IFU and development.